### PR TITLE
Feature add spring damper element

### PR DESCRIPTION
**Pull request by qaumann**
Adding a massless 2-node element which can be used as a relative spring and/or damper. Tests for harmonic excitation, free oscillation for damped and undamped cases as well as an eigenvalue analysis are included.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+        const array_1d<double, 3>& vel = GetGeometry()[i].FastGetSolutionStepValue( VELOCITY, Step );
+        // const array_1d<double, 3>& avel = GetGeometry()[i].FastGetSolutionStepValue( ANGULAR_VELOCITY, Step );
+        unsigned int index = i * 6;
+        rValues[index]   = vel[0];
+        rValues[index+1] = vel[1];
+        rValues[index+2] = vel[2];
+        // rValues[index+3] = avel[0];
+        // rValues[index+4] = avel[1];
+        // rValues[index+5] = avel[2];
+        rValues[index+3] = 0.0;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp:194]: Why don't you use the values of the Angluar Velocity here?
Same for Acceleration

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py [hunk 2]
```diff
@@ -120,15 +121,16 @@ def AssambleTestSuites():
     smallSuite.addTest(TSprismMembranePatchTests('test_execution'))
     smallSuite.addTest(TSprismBendingPatchTests('test_execution'))
     # Shell tests
-    smallSuite.addTest(TShellQ4ThickBendingRollUpTests('test_execution'))
-    smallSuite.addTest(TShellQ4ThickDrillingRollUpTests('test_execution'))
-    smallSuite.addTest(TShellT3ThinBendingRollUpTests('test_execution'))
-    smallSuite.addTest(TShellT3ThinDrillingRollUpTests('test_execution'))
+    # smallSuite.addTest(TShellQ4ThickBendingRollUpTests('test_execution'))
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: Should these test be commented out?

File: applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+
+    KRATOS_CATCH( "" );
+}
+
+//***********************************************************************************
+//***********************************************************************************
+
+void SpringDamperElement3D2N::CalculateRightHandSide(VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo )
+{
+    // std::cout << "CalculateRightHandSide for id=" << this->Id() << std::endl;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp]: I recommend to do some more code cleaning.
You can still recover these by restoring the development branch, should you be needing them later

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py (hunk 2 above)
Comment (qaumann) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: Of course not, I reverted that.

File: applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp (hunk 3 above)
Comment (qaumann) [reply applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp]: Sure, I missed some of those.

File: applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp (hunk 1 above)
Comment (qaumann) [reply applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp:194]: These acceleration values are remains of when I thought about adding mass to the element. As it is massless now, the acceleration should not have an impact on the solution.
For the velocity, I am not sure if there are applications for a rotational damper so I left the angular velocity out for the moment. As the damping matrix is all 0 on the rotation dofs, it neither has an impact on the solution.

File: applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+
+    for ( size_t i = 0; i < 2*OPT_NUM_DIMS; ++i )
+    {
+        rLeftHandSideMatrix( i, i) += elemental_stiffness[i];
+        rLeftHandSideMatrix( i+6, i+6) += elemental_stiffness[i];
+        rLeftHandSideMatrix( i, i+6) -= elemental_stiffness[i];
+        rLeftHandSideMatrix( i+6, i) -= elemental_stiffness[i];
+    }
+
+    // KRATOS_WATCH(rLeftHandSideMatrix);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/spring_damper_element_3D2N.cpp]: found another one :)

**Review by philbucher — APPROVED** [review]
Looks good to me

**Comment by qaumann** [comment]
Thanks for the comments and the approval

**Review by RiccardoRossi — COMMENTED** [review]
hi Quirin,

you should try to avoid accessing by components if you can.

that is, it would be better if you do

array_1d<double,3>& tmp = this->GetValue(NODAL_STIFFNESS)
and then you copy from tmp to elemental_stiffness, as this way you do just one access to the database instead of 2

**Comment by qaumann** [comment]
Hi Riccardo,
thanks for you comment. I will change this when I include the rotational stiffness to the element.

**Pull request merged**